### PR TITLE
tls/ossl: Adjusted value of ERROR_UNEXPECTED_PACKET

### PR DESCRIPTION
**Pull request by michael-redpanda**
Fixes: CORE-8644

Now using SSL_R_WRONG_VERSION_NUMBER, which is the error OpenSSL returns when it receives a non-TLS packet when expecting a TLS handshake packet. This helps Redpanda because it treats ERROR_UNEXPECTED_PACKET as a benign disconnect error.

Added method that can be used to override the OpenSSL message with one that may be more helpful to users.

### CODE REVIEW

File: src/net/ossl.cc [hunk 1]
```diff
@@ -85,7 +85,22 @@ struct is_error_code_enum<seastar::ossl_errc> : true_type {};
 
 template<>
 struct fmt::formatter<seastar::ossl_errc> : public fmt::formatter<std::string_view> {
+    /**
+     * Some OpenSSL errors are hard to parse.  This method can be used to provide
+     * an alternative, more readable, error message.
+     */
+    static std::optional<seastar::sstring> alternate_message(unsigned long error) {
+        if (error == seastar::tls::ERROR_UNEXPECTED_PACKET) {
+            return "Wrong SSL Version number: ensure client is configured to use TLS";
```
[the remark this answers is not in the document]
Comment (michael-redpanda) [reply src/net/ossl.cc:94]: This function is only used when OpenSSL is in use by Seastar

File: src/net/ossl.cc [hunk 2]
```diff
@@ -2068,7 +2083,7 @@ const int seastar::tls::ERROR_PUSH = int(ERR_SYSTEM_FLAG | EPIPE);
 const int seastar::tls::ERROR_PULL = ERR_PACK(
   ERR_LIB_SSL, 0, SSL_R_READ_BIO_NOT_SET);
 const int seastar::tls::ERROR_UNEXPECTED_PACKET = ERR_PACK(
-  ERR_LIB_SSL, 0, SSL_R_UNEXPECTED_MESSAGE);
+  ERR_LIB_SSL, 0, SSL_R_WRONG_VERSION_NUMBER);
```
[the remark this answers is not in the document]
Comment (michael-redpanda) [reply src/net/ossl.cc:2086]: `seastar::tls::ERROR_UNEXPECTED_PACKET`, which when GnuTLS is in use maps to `GNUTLS_E_UNEXPECTED_PACKET`, is what was returned by the GnuTLS library when a non-TLS packet is sent to a TSL endpoint.  This covered a lot of cases.  OpenSSL, however, can return a number of different messages.  The vast majority of time it is `SSL_R_WRONG_VERSION_NUMBER` is what is returned. 

Thinking through this some more, maybe I'll revert this change and add in a few more error codes to the Seastar API to cover more of these conditions.

**Comment by michael-redpanda** [comment]
Force push [f01e249](https://github.com/redpanda-data/seastar/commit/f01e249cd2c83d87d5c22f48b7ab6d3e6d894410):

* Added additional error codes
* Reverted change
* Updated alternate message

File: src/net/ossl.cc [hunk 3]
```diff
@@ -2077,3 +2076,37 @@ const int seastar::tls::ERROR_DECRYPTION_FAILED = ERR_PACK(
...
 const int seastar::tls::ERROR_MAC_VERIFY_FAILED = ERR_PACK(
   ERR_LIB_SSL, 0, SSL_R_DECRYPTION_FAILED_OR_BAD_RECORD_MAC);
+const int seastar::tls::ERROR_WRONG_VERSION_NUMBER = ERR_PACK(
+  ERR_LIB_SSL, 0, SSL_R_WRONG_VERSION_NUMBER);
+const int seastar::tls::ERROR_HTTP_REQUEST = ERR_PACK(
+  ERR_LIB_SSL, 0, SSL_R_HTTP_REQUEST);
+const int seastar::tls::ERROR_HTTPS_PROXY_REQUEST = ERR_PACK(
+  ERR_LIB_SSL, 0, SSL_R_HTTPS_PROXY_REQUEST);
+
+std::optional<seastar::sstring> fmt::formatter<seastar::ossl_errc>::alternate_message(unsigned long error) {
```
Comment (BenPope) [inline src/net/ossl.cc:2086]: It's probably worth accepting an `ossl_errc`, besides, the constants are of type `int`, not `unsigned long`.

File: src/net/ossl.cc [hunk 4]
```diff
@@ -2077,3 +2076,37 @@ const int seastar::tls::ERROR_DECRYPTION_FAILED = ERR_PACK(
...
+    }
+
+    return std::nullopt;
+}
+
+auto fmt::formatter<seastar::ossl_errc>::format(seastar::ossl_errc error, fmt::format_context& ctx) const -> decltype(ctx.out()) {
+    auto alternate = alternate_message(static_cast<unsigned long>(error));
+    if (alternate.has_value()) {
+        return fmt::format_to(ctx.out(), "{}", alternate.value());
+    }
```
Comment (BenPope) [inline src/net/ossl.cc:2103]:
```suggestion
    if (auto alternate = alternate_message(error); alternate.has_value()) {
        return fmt::format_to(ctx.out(), "{}", alternate.value());
    }

**Comment by michael-redpanda** [comment]
Force push [bc055c1](https://github.com/redpanda-data/seastar/commit/bc055c126ec2dd4ef62a2b1b19ef6e8b94210326):

* Updates per PR comments

**Review by BenPope — APPROVED** [review]
:heart:

**Review by aanthony-rp — APPROVED** [review]
